### PR TITLE
[SQL] run time error in integer arithmetic overflow

### PR DESCRIPTION
**Pull request by abhizer**
Is this a user-visible change (yes/no): no

Partially fixes: #1186 

There are still some cases that fail due to the Calcite optimized version performing wrapping arithmetic.

### CODE REVIEW

File: sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/SqlIoTest.java [hunk 1]
```diff
@@ -540,6 +540,43 @@ public void compare(String query, String expected, boolean optimize) {
         this.addRustTestCase(query, compiler, circuit, streams);
     }
 
+    // Unsure what to do with `data`
```
Comment (mihaibudiu) [inline sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/SqlIoTest.java:543]: if the argument is unused it shouldn't exist. The only exception is if you implement an interface which requires the argument.

File: sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/SqlIoTest.java [hunk 2]
```diff
@@ -540,6 +540,43 @@ public void compare(String query, String expected, boolean optimize) {
...
+        InputOutputPair streams = new InputOutputPair(
+                this.getPreparedInputs(compiler),
+                new DBSPZSetLiteral.Contents[] { result }
+        );
+
+        this.addFailingRustTestCase(query, message, compiler, circuit, streams);
+    }
+
+    @Override
+    protected void runtimeFail(String queryAndOutput, String message, InputOutputPair... data) {
```
Comment (mihaibudiu) [inline sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/SqlIoTest.java:564]: Let's use a different function name instead of overriding the existing one.
These are after all different functions.
So create something like qf with no data argument.

File: sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/postgres/PostgresInt8Tests.java [hunk 3]
```diff
@@ -541,33 +541,79 @@ public void testSelect() {
...
 
     @Test
     public void testOutOfRangeCast() {
-        this.runtimeFail("select '-9223372036854775809'::int64", "Overflow", this.getEmptyIOPair());
-        this.runtimeFail("select '9223372036854775808'::int64", "Overflow", this.getEmptyIOPair());
+        String msg = "Overflow";
+
+        this.runtimeFail("""
+            select '-9223372036854775809'::int64;
+             int64
```
Comment (mihaibudiu) [inline sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/postgres/PostgresInt8Tests.java:559]: what's the point in having a table with no rows? The header is ignored anyway.

File: sql-to-dbsp-compiler/lib/sqllib/src/operators.rs [hunk 4]
```diff
@@ -86,19 +86,19 @@ for_all_compare!(gte, bool);
 #[inline(always)]
 fn plus<T>(left: T, right: T) -> T
 where
-    T: Add<Output = T>,
+    T: SQLArithmetic,
```
Comment (mihaibudiu) [inline sql-to-dbsp-compiler/lib/sqllib/src/operators.rs:89]: I think this should be a separate PR

File: sql-to-dbsp-compiler/lib/sqllib/src/sql_arithmetic.rs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    fn minus(&self, other: &Self) -> Self {
+        self - other
+    }
+
+    fn mul(&self, other: &Self) -> Self {
+        self * other
+    }
+}
+
+impl SQLArithmetic for Decimal {
```
Comment (mihaibudiu) [inline sql-to-dbsp-compiler/lib/sqllib/src/sql_arithmetic.rs:97]: This is not sufficient, since it does not check for overflow.
But the Decimal type does not carry enough information to decide whether there is an overflow.
The compiler will have to generate code in a different way, by inserting a cast for the result.
So this is not really a correct implementation the SQL arithmetic.

File: sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/SqlIoTest.java (hunk 2 above)
Comment (mihaibudiu) [reply sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/SqlIoTest.java:564]: This should not have a query and output, just a query.

File: sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/SqlIoTest.java [hunk 6]
```diff
@@ -540,6 +540,43 @@ public void compare(String query, String expected, boolean optimize) {
...
+    private void runtimeFailInner(boolean optimize, String query, String expected, String message, InputOutputPair... _data) {
+        DBSPCompiler compiler = this.testCompiler(optimize);
+        this.prepareData(compiler);
+        compiler.compileStatement("CREATE VIEW VV AS " + query);
+        if (!compiler.options.languageOptions.throwOnError)
+            compiler.throwIfErrorsOccurred();
+        compiler.optimize();
+        DBSPCircuit circuit = getCircuit(compiler);
+        DBSPType outputType = circuit.getOutputType(0);
+        DBSPZSetLiteral.Contents result = this.parseTable(expected, outputType);
```
Comment (mihaibudiu) [inline sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/SqlIoTest.java:554]: there is no expected output, so there should be no parseTable since the table is always empty.
You can just synthesize an empty ZSetLiteral for the output with the outputType.

**Review by mihaibudiu — CHANGES_REQUESTED** [review]

**Comment by mihaibudiu** [comment]
Are you planning to close this PR and open a set of new ones with some of the changes here?

**Comment by abhizer** [comment]
Yes. 
Closing this and breaking it down in multiple PRs.
